### PR TITLE
Add CRUD on own plugin repository

### LINKED ISSUE #33: Adapt the plugin to the new authentication plugin structure

### LINKED ISSUE #40: Add a getById method
We need to be able to get a user from it's "_id" and not only from it's "kuid".
That way we would be able to get a user from something understandable.

### PR DESCRIPTION
**Pull request by jenow**
# Description

This PR is a refactoring according to https://github.com/kuzzleio/kuzzle/pull/804
Depends on https://github.com/kuzzleio/kuzzle/pull/804

# Related issue

* fix #33 
* fix #40 
* fix https://github.com/kuzzleio/kuzzle-private-backlog/issues/22 (private issue)

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -76,57 +77,266 @@ class AuthPassportLocal {
...
+    }
+
+    if (!credentials.username) {
+      return Promise.reject(new this.context.errors.BadRequestError('Username is a mandatory field for authentication strategy "local".'));
     }
-    catch (error) {
-      callback(error);
+
+    if (isUpdate && !credentials.password) {
+      return Promise.reject(new this.context.errors.BadRequestError('Username is a mandatory field for authentication strategy "local".'));
```
Comment (xbill82) [inline lib/index.js]: The condition and the message do not match.

File: lib/index.js [hunk 2]
```diff
@@ -76,57 +77,266 @@ class AuthPassportLocal {
...
-  cleanUserCredentials (user, callback) {
-    try {
-      if (user._source !== undefined) {
-        delete user._source.password;
-      }
-      delete user.password;
-      callback(null, user);
+  validate (request, credentials, kuid, isUpdate) {
+    if (isUpdate && !credentials.username) {
+      return Promise.resolve(true);
```
Comment (xbill82) [inline lib/index.js]: I do not really understand this check: if it's an update and there's no username, we're happy to skip the password check below?

**Comment by dbengsch** [comment]
@xbill82 done

File: lib/index.js [hunk 3]
```diff
@@ -76,57 +77,269 @@ class AuthPassportLocal {
...
+   * @param {object} credentials
+   * @param {string} kuid
+   * @param {boolean} isUpdate
+   * @returns {Promise<boolean>}
    */
-  cleanUserCredentials (user, callback) {
-    try {
-      if (user._source !== undefined) {
-        delete user._source.password;
+  validate (request, credentials, kuid, isUpdate) {
```
Comment (ballinette) [inline lib/index.js]: `validate` signature takes 5 parameters: missing `strategy` between `kuid` and `isUpdate`

File: lib/index.js [hunk 4]
```diff
@@ -76,57 +77,269 @@ class AuthPassportLocal {
...
     }
+
+    return this.getUsersRepository().get(credentials.username)
+      .then(result => {
+        if (result === null) {
+          return Promise.resolve(true);
+        }
+
+        if (kuid !== result.kuid) {
+          return Promise.reject(new this.context.errors.BadRequestError(`Login "${credentials.username}" is already used.`));
```
Comment (ballinette) [inline lib/index.js]: should be a `PreconditionError` instead

File: lib/index.js [hunk 5]
```diff
@@ -76,57 +77,269 @@ class AuthPassportLocal {
...
+   * @returns {Promise<object>}
+   */
+  create (request, credentials, kuid) {
+    if (!credentials.password) {
+      return Promise.reject(new this.context.errors.BadRequestError('Password needed.'));
     }
+    return this.exists(request, kuid)
+      .then(exists => {
+        if (exists) {
+          return Promise.reject(new this.context.errors.BadRequestError('A strategy already exists for this user.'));
```
Comment (ballinette) [inline lib/index.js]: should be a `PreconditionError` instead

File: lib/index.js [hunk 6]
```diff
@@ -76,57 +77,269 @@ class AuthPassportLocal {
...
-      return callback(new this.context.errors.BadRequestError('Missing or invalid given property: password'));
+  /**
+   * @param {KuzzleRequest} request
+   * @param {object} credentials
+   * @param {string} kuid
+   * @returns {Promise<object>}
+   */
+  update (request, credentials, kuid) {
+    if (credentials.kuid) {
+      return Promise.reject(new this.context.errors.BadRequestError('The request must contain a kuid attribute.'));
```
Comment (ballinette) [inline lib/index.js]: "the request must **not** contain a kuid attribute"

File: lib/index.js [hunk 7]
```diff
@@ -76,57 +77,269 @@ class AuthPassportLocal {
...
     }
+    return this.getCredentialsFromUserId(kuid)
+      .then(document => {
+        const credentialUpdate = {};
+
+        if (document === null) {
+          return Promise.reject(new this.context.errors.BadRequestError('A strategy does not exist for this user.'));
```
Comment (ballinette) [inline lib/index.js]: should be a `PreconditionError` instead

File: lib/index.js [hunk 8]
```diff
@@ -76,57 +77,269 @@ class AuthPassportLocal {
...
-    if (request.input.body.password.trim() === '') {
-      return callback(new this.context.errors.BadRequestError('Empty password is not allowed'));
+        if (credentials.password) {
+          credentialUpdate.userPassword = this.passwordManager.encryptPassword(credentials.password);
+        }
+
+        // To change the username, we have to create a new (complete) document and remove the old one
+        if (credentials.username && document._id !== credentials.username) {
+          let createdDocument;
+          const oldDocumentId = document._id;
```
Comment (ballinette) [inline lib/index.js]: unnecessary. we can only calll `return this.getUsersRepository().delete(document._id);` at line 244

File: lib/index.js [hunk 9]
```diff
@@ -76,57 +77,269 @@ class AuthPassportLocal {
...
+  /**
+   * @param {KuzzleRequest} request
+   * @param {string} kuid
+   * @returns {Promise<object>}
+   */
+  delete (request, kuid) {
+    return this.getCredentialsFromUserId(kuid)
+      .then(document => {
+        if (document === null) {
+          return Promise.reject(new this.context.errors.BadRequestError('A strategy does not exist for this user.'));
```
Comment (ballinette) [inline lib/index.js]: should be a `PreconditionError` instead

File: lib/index.js [hunk 10]
```diff
@@ -76,57 +77,269 @@ class AuthPassportLocal {
...
+  /**
+   * @param {KuzzleRequest} request
+   * @param {string} username
+   * @returns {Promise<object>}
+   */
+  getById (request, username) {
+    return this.getUsersRepository().get(username)
+      .then(document => {
+        if (document === null) {
+          return Promise.reject(new this.context.errors.BadRequestError('A strategy does not exist for this username.'));
```
Comment (ballinette) [inline lib/index.js]: should be a `PreconditionError` instead

File: lib/index.js [hunk 11]
```diff
@@ -76,57 +77,269 @@ class AuthPassportLocal {
...
+  /**
+   * @param {KuzzleRequest} request
+   * @param {string} kuid
+   * @returns {Promise<object>}
+   */
+  getInfo (request, kuid) {
+    return this.getCredentialsFromUserId(kuid)
+      .then(document => {
+        if (document === null) {
+          return Promise.reject(new this.context.errors.BadRequestError('A strategy does not exist for this user.'));
```
Comment (ballinette) [inline lib/index.js]: should be a `PreconditionError` instead

File: lib/index.js (hunk 8 above)
Comment (dbengsch) [inline lib/index.js]: I don't understand this comment

Comment (ballinette) [inline lib/index.js]: I mean : we don't need to assign a variable `oldDocumentId`.

and we can replace line 242 below (`return this.getUsersRepository().delete(oldDocumentId);`) by:
```javascript
return this.getUsersRepository().delete(document._id);
```

**Pull request merged**